openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/doc/man/man8/strvf.8.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) strvf.8.man,v 0.9.2.17 2006/04/05 09:43:51 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:43:51 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" strvf.8.man,v
.\" Revision 0.9.2.17  2006/04/05 09:43:51  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database streams.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so streams.macros
.\"
.\"
.TH STRVF 8 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Administration"
.\"
.\"
.SH NAME
.B strvf
\- \fISTREAMS\fR verification tool
.\"
.\"
.SH SYNOPSIS
.PP
.HP 12
\fBstrvf\fR [\fIoptions\fR]
.PD 0
.HP 12
\fBstrvf\fR {\fB-h\fR|\fB--help\fR}
.HP 12
\fBstrvf\fR {\fB-V\fR|\fB--version\fR}
.HP 12
\fBstrvf\fR {\fB-C\fR|\fB--copying\fR}
.PD
.\"
.\"
.SH DESCRIPTION
.PP
The
.B strvf
command executes a series of subcommands that verify whether or not
.I STREAMS
is currently installed and configured on your system.  All output is sent to
.IR stdout .
Verbose output is always sent to the log file
.BR /var/log/streams/strvf.log .
.\".BR /var/adm/streams/strvf.log .
.PP
These subcommands make sure that the
.I STREAMS
kernel daemons are running and that
.BR open (2s),
.BR putmsg (2),
.BR getmsg (2),
.BR ioctl (2s)
and
.BR close (2)
can be performed on
.BR /dev/echo .
.\"
.\"
.SH OPTIONS
.PP
.TP
\fB-n\fR, \fB--nologging\fR
Suppress logging, send output to
.B stdout
and errors to
.BR stderr .
Default: off.
.TP
\fB-b\fR, \fB--basename\fR \fIBASENAME\fR
Only significant when logging.
Set the base name of logging and error files.  Default:
.RB \(lq strvf \(rq.
.TP
\fB-O\fR, \fB--directory\fR \fIDIRECTORY\fR
Only significant when logging.
Set the directory to which log and error files are written.  Default:
.RB \(lq /var/log/streams \(rq.
.TP
\fB-o\fR, \fB--outfile\fR \fIOUTFILE\fR
Only significant when logging.
Set the output file name for logging.  Default:
\(lq\fIBASENAME\fB.log\fR\(rq.
.TP
\fB-e\fR, \fB--errfile\fR \fIERRFILE\fR
Only significant when logging.
Set the output file name for errors.  Default:
\(lq\fIBASENAME\fB.err\fR\(rq.
.TP
\fB-D\fR, \fB--devname\fR \fIDEVNAME\fR
Only significant when logging.
Set the device name to perform test upon.  Default:
.RB \(lq /dev/echo \(rq.
.TP
\fB-q\fR, \fB--quiet\fR
Suppress normal output.
.TP
\fB-d\fR, \fB--debug\fR [\fIlevel\fR]
Increase, or specify the debugging of the output.
.TP
\fB-v\fR, \fB--verbose\fR [\fIlevel\fR]
Increase, or specify the verbosity of the output.
.TP
\fB-h\fR, \fB--help\fR
Print usage information and exit.
.TP
\fB-V\fR, \fB--version\fR
Print version information and exit.
.TP
\fB-C\fR, \fB--copying\fR
Print copying information and exit.
.\"
.\"
.SH DIAGNOSTICS
.PP
.B strvf
returns a zero (0) exit status on success.  If
.B strvf
fails, it returns an exit status of one (1) and prints a diagnostic error
message to
.BR stderr .
.PP
An exit status of zero (0) indicates that
.B strvf
was successful; one (1) indicates that an error occurred and diagnostic message
is printed on
.BR stderr ;
two (2) indicates that the option syntax was in error and a diagnostic message
is printed on
.BR stderr .
.PP
Errors may result from the failure of any of the test in the verification
procedure, failure to open the
.B /dev/echo
device, or inadequate permissions.
.\"
.\"
.SH NOTICES
.PP
.\"
.\"
.SH EXAMPLES
.PP
Sample output:
.RS
.nf
\fC\s-1
.\"-> Logging results to /var/adm/streams/strvf.log
-> Logging results to /var/log/streams/strvf.log
--------------------------------------------------
-> Verify Linux Fast-STREAMS installation.
-> Verify open, putmsg, getmsg, ioctl, and close
-> can be performed on a stream.
--------------------------------------------------
--------------------------------------------------
-> Linux Fast-STREAMS is installed and operational
--------------------------------------------------
\s+1\fP
.fi
.RE
.\"
.\"
.SH DEVICES
.PP
.B /dev/echo
Loopback
.I STREAMS
driver used by
.BR strvf .
.\"
.\"
.SH FILES
.PP
.TP
.B /usr/sbin/strvf
contains the
.B strvf
command.
.TP
.B /var/log/streams/strvf.log
.\".B /var/adm/streams/strvf.log
contains a verbose description and status of all sub-commands.
.TP
.\"
.\"
.SH "SEE ALSO"
.PP
.BR echo (4),
.BR strtst (8).
.\"
.\"
.SH BUGS
.PP
.B strvf
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.IR HP-UX \(rg.
This is an
.IR HP-UX \(rg-specific
command.
.[
hpuxspg
.]
.B strvf
is provided in
.I @PACKAGE_TITLE@
for compatibility with
.IR HP-UX \(rg
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.IR HP-UX \(rg
does not document any long options.
.IP \(em
.IR HP-UX \(rg
does not document the
.BR -d ", " -q ", " -v ", " -h ", " -V " or " -C options.
.IP \(em
Options
.BR --help " and " --version
are provided for compatibility with GNU coding standards;
.BR --copying ,
OpenSS7 coding standards.
.IP \(em
The format of the output from the
.B strvf
command may be different from that of
.IR HP-UX \(rg.
.\"
.\"
.SH CONFORMANCE
.PP
.IR HP-UX \(rg
documentation.
.[
hpuxdoc
.]
.\"
.\"
.SH HISTORY
.PP
.B strvf
first appeared in
.IR HP-UX \(rg.
.\"
.\"
.[
$LIST$
.]
.TI


openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/src/util/strvf.c
/*****************************************************************************

 @(#) strvf.c,v OpenSS7-0_9_2_D_rc2(0.9.2.17) 2006/03/10 07:23:59

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; version 2 of the License.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2006/03/10 07:23:59 by brian

 *****************************************************************************/

#ident "@(#) strvf.c,v OpenSS7-0_9_2_D_rc2(0.9.2.17) 2006/03/10 07:23:59"

static char const ident[] =
    "strvf.c,v OpenSS7-0_9_2_D_rc2(0.9.2.17) 2006/03/10 07:23:59";

#define _XOPEN_SOURCE 600

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <memory.h>
#include <fcntl.h>
#include <errno.h>
#include <limits.h>

#ifdef _GNU_SOURCE
#include <getopt.h>
#endif

#include <stropts.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/sysmacros.h>
#include <sys/poll.h>
#include <sys/time.h>
#include <sys/ioctl.h>
#include <time.h>
#include <signal.h>
#include <syslog.h>
#include <sys/utsname.h>
#include <sys/strlog.h>

static int logging = 1;
static int debug = 0;			/* default no debug */
static int output = 1;			/* default normal output */

char outfile[256] = "";
char errfile[256] = "";
char outpath[256] = "";
char errpath[256] = "";
char basname[256] = "";
char outpdir[256] = "/var/log/streams";
char devname[256] = "";

static void
version(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	fprintf(stdout, "\
%2$s\n\
Copyright (c) 2001-2006  OpenSS7 Corporation.  All Rights Reserved.\n\
Distributed under GPL Version 2, included here by reference.\n\
See `%1$s --copying' for copying permissions.\n\
", argv[0], ident);
}

static void
usage(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	fprintf(stderr, "\
Usage:\n\
    %1$s [options]\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
", argv[0]);
}

static void
help(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	fprintf(stdout, "\
Usage:\n\
    %1$s [options]\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
Options:\n\
    -n, --nologging\n\
        do not log, send output to stdout and stderr\n\
    -b, --basename BASENAME\n\
        file basename, default: 'strvf'\n\
    -O, --directory DIRECTORY\n\
        redirect output to files in DIRECTORY, default: /var/log/streams\n\
    -o, --outfile OUTFILE\n\
        redirect output to OUTFILE, default: ${BASENAME}.log\n\
    -e, --errfile ERRFILE\n\
        redirect errors to ERRFILE, default: ${BASENAME}.err\n\
    -D, --devname DEVNAME\n\
        test device to open, default: '/dev/echo'\n\
    -q, --quiet\n\
        suppress output\n\
    -d, --debug [LEVEL]\n\
        increment or set debug LEVEL (default: 0)\n\
    -v, --verbose [LEVEL]\n\
        increment or set output verbosity LEVEL (default: 1)\n\
    -h, --help, -?\n\
        print this usage information and exit\n\
    -V, --version\n\
        print the version and exit\n\
    -C, --copying\n\
        print copying permissions and exit\n\
", argv[0]);
}

static void
copying(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	fprintf(stdout, "\
--------------------------------------------------------------------------------\n\
%1$s\n\
--------------------------------------------------------------------------------\n\
Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com>\n\
Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>\n\
\n\
All Rights Reserved.\n\
--------------------------------------------------------------------------------\n\
This program is free software; you can  redistribute  it and/or modify  it under\n\
the terms  of the GNU General Public License  as  published by the Free Software\n\
Foundation; either version 2 of the License.\n\
\n\
This program is distributed in the hope that it will  be useful, but WITHOUT ANY\n\
WARRANTY; without even  the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU General Public License for more details.\n\
\n\
You should  have received a copy of the GNU  General  Public License  along with\n\
this program; if not, write to the Free Software Foundation, Inc., 675 Mass Ave,\n\
Cambridge, MA 02139, USA.\n\
--------------------------------------------------------------------------------\n\
U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on behalf\n\
of the U.S. Government (\"Government\"), the following provisions apply to you. If\n\
the Software is supplied by the  Department of Defense (\"DoD\"), it is classified\n\
as \"Commercial  Computer  Software\"  under  paragraph  252.227-7014  of the  DoD\n\
Supplement  to the  Federal Acquisition Regulations  (\"DFARS\") (or any successor\n\
regulations) and the  Government  is acquiring  only the  license rights granted\n\
herein (the license rights customarily provided to non-Government users). If the\n\
Software is supplied to any unit or agency of the Government  other than DoD, it\n\
is  classified as  \"Restricted Computer Software\" and the Government's rights in\n\
the Software  are defined  in  paragraph 52.227-19  of the  Federal  Acquisition\n\
Regulations (\"FAR\")  (or any successor regulations) or, in the cases of NASA, in\n\
paragraph  18.52.227-86 of  the  NASA  Supplement  to the FAR (or any  successor\n\
regulations).\n\
--------------------------------------------------------------------------------\n\
Commercial  licensing  and  support of this  software is  available from OpenSS7\n\
Corporation at a fee.  See http://www.openss7.com/\n\
--------------------------------------------------------------------------------\n\
", ident);
}

static void
strvf_exit(int retval)
{
	if (retval) {
		fprintf(stderr, "--------------------------------------------------\n");
		fprintf(stderr, "-> Linux Fast-STREAMS is NOT operational.\n");
		fprintf(stderr, "--------------------------------------------------\n");
	} else {
		fprintf(stderr, "--------------------------------------------------\n");
		fprintf(stderr, "-> Linux Fast-STREAMS is installed and operational\n");
		fprintf(stderr, "--------------------------------------------------\n");
	}
	fflush(stderr);
	exit(retval);
}

static void
strvf(int argc, char *argv[])
{
	int fd;

	if (basname[0] == '\0')
		snprintf(basname, sizeof(basname), "strvf");
	if (logging || outfile[0] != '\0') {
		if (outfile[0] == '\0')
			snprintf(outfile, sizeof(outfile), "%s.log", basname);
		snprintf(outpath, sizeof(outpath), "%s/%s", outpdir, outfile);
		if (output > 1)
			syslog(LOG_NOTICE, "Redirecting stdout to file %s", outpath);
		fflush(stdout);
		if (freopen(outpath, "a", stderr) == NULL) {
			syslog(LOG_ERR, "%m");
			syslog(LOG_ERR, "Could not redirect stdout to %s", outpath);
			exit(1);
		}
	}
	if (logging || errfile[0] != '\0') {
		if (errfile[0] == '\0')
			snprintf(errfile, sizeof(errfile), "%s.err", basname);
		snprintf(errpath, sizeof(errpath), "%s/%s", outpdir, errfile);
		if (output > 1)
			syslog(LOG_NOTICE, "Redirecting stderr to file %s", errpath);
		fflush(stderr);
		if (freopen(errpath, "a", stderr) == NULL) {
			syslog(LOG_ERR, "%m");
			syslog(LOG_ERR, "Could not redirect stderr to %s", errpath);
			exit(1);
		}
	}
	syslog(LOG_NOTICE, "Startup complete.");
	fprintf(stderr, "-> Logging results to %s\n", outpath);
	fprintf(stderr, "--------------------------------------------------\n");
	fprintf(stderr, "-> Verify Linux Fast-STREAMS installation.\n");
	fprintf(stderr, "-> Verify open, putmsg, getmsg, ioctl, and close\n");
	fprintf(stderr, "-> can be performed on a stream.\n");
	fprintf(stderr, "--------------------------------------------------\n");
	fflush(stderr);
	{
		fprintf(stdout, "Open device %s\n", devname);
		fflush(stdout);
		if ((fd = open(devname, O_RDWR)) < 0) {
			fprintf(stdout, "Open failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Open successful.\n");
		fflush(stdout);
	}
	{
		int ret, flags = 0;
		char ctlbuf[] = "\0\1\2\3\4\5\6\7";
		char datbuf[] = "\0\1\2\3\4\5\6\7";
		struct strbuf ctl = { sizeof(ctlbuf), sizeof(ctlbuf), ctlbuf };
		struct strbuf dat = { sizeof(datbuf), sizeof(datbuf), datbuf };

		fprintf(stdout, "Putmsg to device %s\n", devname);
		fflush(stdout);
		if ((ret = putmsg(fd, &ctl, &dat, flags)) < 0) {
			fprintf(stdout, "Putmsg failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Putmsg successful.\n");
		fflush(stdout);
	}
	{
		int ret, flags = 0;
		char ctlbuf[8];
		char datbuf[8];
		struct strbuf ctl = { sizeof(ctlbuf), sizeof(ctlbuf), ctlbuf };
		struct strbuf dat = { sizeof(datbuf), sizeof(datbuf), datbuf };

		fprintf(stdout, "Getmsg from device %s\n", devname);
		fflush(stdout);
		if ((ret = getmsg(fd, &ctl, &dat, &flags)) < 0) {
			fprintf(stdout, "Getmsg failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Getmsg successful.\n");
		fflush(stdout);
	}
	{
		int ret;
		char iocbuf[] = "\0\1\2\3\4\5\6\7";
		struct strioctl ioc = { I_STR, 0, sizeof(iocbuf), iocbuf };

		fprintf(stdout, "Ioctl of device %s\n", devname);
		fflush(stdout);
		if ((ret = ioctl(fd, I_STR, &ioc)) < 0) {
			fprintf(stdout, "Ioctl failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Ioctl successful.\n");
		fflush(stdout);
	}
	{
		fprintf(stdout, "Close device %s\n", devname);
		fflush(stdout);
		if (close(fd) < 0) {
			fprintf(stdout, "Close failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Close successful.\n");
		fflush(stdout);
	}
	strvf_exit(0);
}

int
main(int argc, char *argv[])
{
	for (;;) {
		int c, val;

#if defined _GNU_SOURCE
		int option_index = 0;
		/* *INDENT-OFF* */
		static struct option long_options[] = {
			{"nologging",	no_argument,		NULL, 'n'},
			{"basename",	required_argument,	NULL, 'b'},
			{"directory",	required_argument,	NULL, 'O'},
			{"outfile",	required_argument,	NULL, 'o'},
			{"errfile",	required_argument,	NULL, 'e'},
			{"devname",	required_argument,	NULL, 'D'},
			{"quiet",	no_argument,		NULL, 'q'},
			{"debug",	optional_argument,	NULL, 'd'},
			{"verbose",	optional_argument,	NULL, 'v'},
			{"help",	no_argument,		NULL, 'h'},
			{"version",	no_argument,		NULL, 'V'},
			{"copying",	no_argument,		NULL, 'C'},
			{"?",		no_argument,		NULL, 'H'},
			{ 0, }
		};
		/* *INDENT-ON* */

		c = getopt_long_only(argc, argv, "n:b:O:o:e:D:qd::v::hVC?W:", long_options,
				     &option_index);
#else				/* defined _GNU_SOURCE */
		c = getopt(argc, argv, "n:b:O:o:e:D:qdvhVC?");
#endif				/* defined _GNU_SOURCE */
		if (c == -1) {
			if (debug)
				fprintf(stderr, "%s: done options processing\n", argv[0]);
			break;
		}
		switch (c) {
		case 0:
			goto bad_usage;
		case 'n':	/* -n, --nologging */
			logging = 0;
			break;
		case 'b':
			strncpy(basname, optarg, sizeof(basname));
			break;
		case 'O':
			strncpy(outpdir, optarg, sizeof(outpdir));
			break;
		case 'o':
			strncpy(outfile, optarg, sizeof(outfile));
			break;
		case 'e':
			strncpy(errfile, optarg, sizeof(errfile));
			break;
		case 'D':
			strncpy(devname, optarg, sizeof(devname));
			break;
		case 'd':	/* -d, --debug [LEVEL] */
			if (debug)
				fprintf(stderr, "%s: increasing debug verbosity\n", argv[0]);
			if (optarg == NULL) {
				debug++;
			} else {
				if ((val = strtol(optarg, NULL, 0)) < 0)
					goto bad_option;
				debug = val;
			}
			break;
		case 'q':	/* -q, --quiet */
			if (debug)
				fprintf(stderr, "%s: suppressing normal output\n", argv[0]);
			debug = 0;
			output = 0;
			break;
		case 'v':	/* -v, --verbose [level] */
			if (debug)
				fprintf(stderr, "%s: increasing output verbosity\n", argv[0]);
			if (optarg == NULL) {
				output++;
				break;
			}
			if ((val = strtol(optarg, NULL, 0)) < 0)
				goto bad_option;
			output = val;
			break;
		case 'h':	/* -h, --help */
		case 'H':	/* -H, --? */
			if (debug)
				fprintf(stderr, "%s: printing help message\n", argv[0]);
			help(argc, argv);
			exit(0);
		case 'V':	/* -V, --version */
			if (debug)
				fprintf(stderr, "%s: printing version message\n", argv[0]);
			version(argc, argv);
			exit(0);
		case 'C':	/* -C, --copying */
			if (debug)
				fprintf(stderr, "%s: printing copying message\n", argv[0]);
			copying(argc, argv);
			exit(0);
		case '?':
		default:
		      bad_option:
			optind--;
		      bad_nonopt:
			if (output || debug) {
				if (optind < argc) {
					fprintf(stderr, "%s: syntax error near '", argv[0]);
					while (optind < argc)
						fprintf(stderr, "%s ", argv[optind++]);
					fprintf(stderr, "'\n");
				} else {
					fprintf(stderr, "%s: missing option or argument", argv[0]);
					fprintf(stderr, "\n");
				}
				fflush(stderr);
			      bad_usage:
				usage(argc, argv);
			}
			exit(2);
		}
	}
	/* 
	 * dont' ignore non-option arguments
	 */
	if (optind < argc)
		goto bad_nonopt;
	strvf(argc, argv);
	exit(0);
}
